### PR TITLE
Migrations with CLI and without knexfile

### LINKED ISSUE #2819: Create migrations with the CLI without knexfile
### Feature discussion / request

It would be convenient to be able to create migrations without knexfile, one could pass the required arguments to the CLI.

In my case only `client` or `dialect` is necessary but one can also imagine passing the directory of the migrations or other by the arguments.

```
knex migrate:make create_rule_table --client mysql --directory ./migrations
```

### PR DESCRIPTION
**Pull request by aurium**
Allows to create and run migrations.

closes #2819

### CODE REVIEW

File: bin/cli.js [hunk 1]
```diff
@@ -35,11 +35,29 @@ function checkLocalModule(env) {
...
+  };
+}
+
+function initKnex(env, commander) {
   checkLocalModule(env);
 
-  if (!env.configPath) {
-    exit('No knexfile found in this directory. Specify a path with --knexfile');
+  if (!env.configuration) {
+    if (commander.client) mkConfigObj(env, commander);
```
Comment (kibertoad) [inline bin/cli.js]: did you run prettier formatting before commit?..

File: bin/cli.js [hunk 2]
```diff
@@ -52,7 +70,10 @@ function initKnex(env) {
 
   var environment = commander.env || process.env.NODE_ENV;
   var defaultEnv = 'development';
-  var config = require(env.configPath);
+  var config =
```
Comment (kibertoad) [inline bin/cli.js]: Would you consider replacing all vars in this file with let/const?

File: bin/cli.js [hunk 3]
```diff
@@ -52,7 +70,10 @@ function initKnex(env) {
...
+    typeof env.configuration == 'object'
```
Comment (kibertoad) [inline bin/cli.js]: since we depend on Lodash anyway, you could just use _.isObject

File: bin/cli.js [hunk 4]
```diff
@@ -35,11 +35,29 @@ function checkLocalModule(env) {
   }
 }
 
-function initKnex(env) {
+function mkConfigObj(env, commander) {
+  env.configuration = {
+    ext: 'js',
+    [commander.env || process.env.NODE_ENV || 'development']: {
```
Comment (kibertoad) [inline bin/cli.js]: this is very difficult to read. Could you please split it somewhat? Inline condition inside key name resolution is heavy :)

File: bin/cli.js [hunk 5]
```diff
@@ -35,11 +35,29 @@ function checkLocalModule(env) {
   }
 }
 
-function initKnex(env) {
+function mkConfigObj(env, commander) {
+  env.configuration = {
```
Comment (kibertoad) [inline bin/cli.js]: so this is mutating passed params? Can this be done in a more immutable way?

File: bin/cli.js [hunk 6]
```diff
@@ -90,6 +112,12 @@ function invoke(env) {
     .option('--debug', 'Run with debugging.')
     .option('--knexfile [path]', 'Specify the knexfile path.')
     .option('--cwd [path]', 'Specify the working directory.')
+    .option('--client [name]', 'Set DB client whitout a knexfile.')
```
Comment (kibertoad) [inline bin/cli.js]: "whitout " -> typo

File: bin/cli.js [hunk 7]
```diff
@@ -90,6 +112,12 @@ function invoke(env) {
...
+    .option('--connection [address]', 'Set DB connection whitout a knexfile.')
```
Comment (kibertoad) [inline bin/cli.js]: same

File: bin/cli.js [hunk 8]
```diff
@@ -136,8 +164,12 @@ function invoke(env) {
       'Specify the stub extension (default js)'
     )
     .action(function(name) {
-      var instance = initKnex(env);
-      var ext = (argv.x || env.configPath.split('.').pop()).toLowerCase();
+      var instance = initKnex(env, commander);
```
Comment (kibertoad) [inline bin/cli.js]: do we need to pass commander instance for internal calls? can't we pass resolved params instead?

File: bin/cli.js [hunk 9]
```diff
@@ -136,8 +164,12 @@ function invoke(env) {
...
+      var ext = (
```
Comment (kibertoad) [inline bin/cli.js]: might make sense to extract to internal function with self-explanatory name or at least add comment with explanation

File: test/migrate-cli.sh [hunk 10]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash -e
+
```
Comment (kibertoad) [inline test/migrate-cli.sh]: wouldn't it be simpler to just write a Jake (https://www.npmjs.com/package/jake) script within usual unit test?

**Comment by kibertoad** [comment]
@aurium Thank you for your contribution!

**Comment by kibertoad** [comment]
CI is failing with "test/migrate-cli.sh: line 4: realpath: command not found" error, btw.

File: bin/cli.js (hunk 1 above)
Comment (aurium) [reply bin/cli.js]: No. Is it necessary?

File: bin/cli.js (hunk 3 above)
Comment (aurium) [reply bin/cli.js]: Thanks! Changing...

File: bin/cli.js (hunk 2 above)
Comment (aurium) [reply bin/cli.js]: I could, however there are many "var's" in the same scope and upper. I could do this in another PR to isolate the changes.

File: bin/cli.js (hunk 8 above)
Comment (aurium) [reply bin/cli.js]: There is a method to do that on commander API? Can't find.

File: bin/cli.js (hunk 9 above)
Comment (aurium) [reply bin/cli.js]: Now using `initKnex(env, commander.opts())`. looks more explicit on its intention.

File: bin/cli.js (hunk 2 above)
Comment (kibertoad) [reply bin/cli.js]: Fair enough :)

**Comment by kibertoad** [comment]
@aurium Tests are failing on node 6 and 8 now. I think "fs.promises" is only available since Node 10, so probably you would have to promisify fs methods manually (e. g. using bluebird).

**Comment by kibertoad** [comment]
LGTM! @elhigu, what do you think?

**Comment by elhigu** [comment]
Hey... awesome client tests! 🎉 Now we can start finishing all the other client PRs which were deferred because of missing tests 😊

**Comment by elhigu** [comment]
And great work everyone!

**Pull request merged**